### PR TITLE
Census view optimizations

### PR DESCRIPTION
**Pull request by thesiv**

### CODE REVIEW

File: census_view.cpp [hunk 1]
```diff
@@ -1268,8 +1268,44 @@ void CensusView::apply_changes
...
+    auto const& all_headers(case_parms()[0].member_names());
+    if(all_headers != all_headers_)
+        {
+        // The entire set of columns has changed, we have no choice but to
+        // reexamine all of them to check which ones should be shown.
+        all_headers_ = all_headers;
+        headers_show_.resize(all_headers_.size(), false);
+        must_update_columns = true;
+        }
+
```
Comment (vadz) [inline census_view.cpp]: I realize this is my own code, but I think there should be `else` here as it doesn't make sense to do anything else if `must_update_columns` is already `true`.

Comment (vadz) [reply census_view.cpp]: > 
> 
> I realize this is my own code, but I think there should be `else` here as it doesn't make sense to do anything else if `must_update_columns` is already `true`.

Actually it's even worse than this: if `all_headers != all_headers_` there doesn't seem to be any reasons for `headers_show` to have as many elements as `all_headers`, so we could just crash below...

File: any_member_test.cpp [hunk 2]
```diff
@@ -103,7 +103,7 @@ std::ostream& operator<<(std::ostream& os, X const& x) {return os << x.str() <<
 class S
     :public MemberSymbolTable<S>
 {
-    friend class any_member_test;
+    friend struct any_member_test;
```
Comment (vadz) [inline any_member_test.cpp]: Please remove this commit from this PR, it's unrelated and I shouldn't have included it in the original one. TIA!

File: any_member.hpp [hunk 3]
```diff
@@ -533,8 +533,7 @@ MemberType const* member_cast(any_member<ClassType> const& member)
 template<typename ClassType>
 class MemberSymbolTable
 {
-    typedef std::map<std::string, any_member<ClassType>> member_map_type;
-    typedef typename member_map_type::value_type member_pair_type;
+    typedef std::vector<any_member<ClassType> > values_vector_type;
```
Comment (vadz) [inline any_member.hpp]:
```suggestion
    using values_vector_type = std::vector<any_member<ClassType>>;
```

File: any_member.hpp [hunk 4]
```diff
@@ -596,25 +595,20 @@ any_member<ClassType>& MemberSymbolTable<ClassType>::operator[]
     (std::string const& s
     )
 {
-    auto i = map_.find(s);
-    if(map_.end() == i)
+    auto i = std::lower_bound(member_names_.begin(), member_names_.end(), s);
```
Comment (vadz) [inline any_member.hpp]: Not sure why did I use `lower_bound()` here instead of `find()`... If you don't see any reason to do it neither, please replace it with the simpler `find()`.

File: any_member.hpp [hunk 5]
```diff
@@ -639,10 +633,23 @@ void MemberSymbolTable<ClassType>::ascribe
...
 
-    ClassType* class_object = static_cast<ClassType*>(this);
-    map_.insert(member_pair_type(s, any_member<ClassType>(class_object, p2m)));
-    // TODO ?? This would appear to be O(N^2).
     auto i = std::lower_bound(member_names_.begin(), member_names_.end(), s);
+
+    ClassType* class_object = static_cast<ClassType*>(this);
+
+    // Using the usual insert() method relies on vector elements being
+    // assignable in C++98, but this is not the case for any_member<>, so we
```
Comment (vadz) [inline any_member.hpp]: This should work just fine in C++17 however, if `any_member` is movable which, AFAICS, it should be. Could you please make it so and get rid of using `swap()` here?

File: any_member.hpp [hunk 6]
```diff
@@ -545,6 +545,9 @@ class MemberSymbolTable
     bool equals(MemberSymbolTable<ClassType> const&) const;
     std::vector<std::string> const& member_names() const;
 
+    size_t get_member_index(std::string const&) const;
```
Comment (vadz) [inline any_member.hpp]: We probably should use `int` here now as lmi moved away from using `size_t` since the original PR was done.

File: census_view.cpp [hunk 7]
```diff
@@ -1075,8 +1075,24 @@ Input* CensusView::class_parms_from_class_name(std::string const& class_name)
 bool CensusView::column_value_varies_across_cells(std::string const& header) const
 {
     auto const z = case_parms()[0][header];
-    for(auto const& j : class_parms()) {if(z != j[header]) return true;}
-    for(auto const& j : cell_parms() ) {if(z != j[header]) return true;}
+
+    size_t const class_header_index = class_parms().begin()->get_member_index(header);
```
Comment (vadz) [inline census_view.cpp]: Can we be sure that `class_parms()` here (and `cell_parms()` below) are not empty?

Comment (thesiv) [reply census_view.cpp]: Looks like we can:
```cpp
multiple_cell_document::multiple_cell_document()
    :case_parms_  (1)
    ,class_parms_ (1)
    ,cell_parms_  (1)
{
    assert_vector_sizes_are_sane();
}
...
multiple_cell_document::multiple_cell_document(std::string const& filename)
{
    xml_lmi::dom_parser parser(filename);
    parse(parser);
}
...
void multiple_cell_document::parse(xml_lmi::dom_parser const& parser)
{
...
    assert_vector_sizes_are_sane();
}
...
void multiple_cell_document::assert_vector_sizes_are_sane() const
{
    LMI_ASSERT(1 == case_parms_.size());
    LMI_ASSERT(    !class_parms_.empty());
    LMI_ASSERT(    !cell_parms_ .empty());
}
```

**Comment by thesiv** [comment]
Fixed.

File: any_member.hpp [hunk 8]
```diff
@@ -323,6 +331,17 @@ any_member<ClassType>& any_member<ClassType>::operator=(std::string const& s)
     return assign(s);
 }
 
+template<typename ClassType>
+any_member<ClassType>& any_member<ClassType>::operator=
+    (any_member<ClassType>&& other
+    )
+{
+    LMI_ASSERT(other.content_);
```
Comment (vadz) [inline any_member.hpp]: I see that this is done in the assignment operator as well, but this doesn't really make sense to me. Do you understand why should it be invalid to move from an invalid/default-initialized `any_member`? If not, I'd rather leave this assert out.

File: any_member.hpp [hunk 9]
```diff
@@ -400,6 +400,16 @@ any_member<ClassType>& any_member<ClassType>::assign(std::string const& s)
     return *this;
 }
 
+/// The standard implementation of swap() doesn't work for any_member<> as it
```
Comment (vadz) [inline any_member.hpp]: I'm not sure if this is still true in C++11.

Also, I don't think we actually need a global `swap()` any more, do we? I.e. this commit could be simple dropped.

Finally, if you decide to keep it for some reason, there is a typo in the commit message:

> hence we this swap() function()

should be just

> hence this swap() function

File: census_view.hpp [hunk 10]
```diff
@@ -134,6 +134,14 @@ class CensusView final
     wxDataViewCtrl* list_window_;
     wxObjectDataPtr<CensusViewDataViewModel> list_model_;
 
+    struct header_cache
+    {
+        std::string name;
+        bool        show;
```
Comment (vadz) [inline census_view.hpp]: We have `oenum_visibility` now and it would be better to use it here, i.e. make this `oenum_visibility visibility{oe_shown};` (I've added an initializer because there doesn't seem to be any good reason to not initialize this field...).

File: census_view.hpp [hunk 11]
```diff
@@ -134,6 +134,14 @@ class CensusView final
     wxDataViewCtrl* list_window_;
     wxObjectDataPtr<CensusViewDataViewModel> list_model_;
 
+    struct header_cache
```
Comment (vadz) [inline census_view.hpp]: Sorry for nitpicking, but I don't think this is a good name because this is not really a cache at all. I'd call it just `header_column` because it stores information about a column.

File: census_view.cpp [hunk 12]
```diff
@@ -1292,20 +1341,18 @@ void CensusView::update_visible_columns()
     // Reason: although the case and class defaults are hidden, they're
     // still information--so if the user made them different from any cell
     // wrt some column, we respect that conscious decision.
-    std::vector<std::string> const& all_headers(case_parms()[0].member_names());
-    int column = 0;
-    for(auto const& header : all_headers)
+    int column = 1;
```
Comment (vadz) [inline census_view.cpp]: Why the change from initializing it to 0 and incrementing it in the beginning of the loop? IMHO the old/current version is more natural, initializing to 1 seems strange.

File: census_view.cpp [hunk 13]
```diff
@@ -1292,20 +1341,18 @@ void CensusView::update_visible_columns()
...
+    for(auto const& header : all_headers_)
         {
-        ++column;
-        if(column_value_varies_across_cells(header))
+        if(header.show)
```
Comment (vadz) [inline census_view.cpp]: With `oenum_visibility` change, this should become `switch(header.visibility)`.

File: census_view.cpp [hunk 14]
```diff
@@ -1268,8 +1268,57 @@ void CensusView::apply_changes
...
-{
+bool CensusView::update_visible_columns()
+{
+    // Check if we really need to update the columns, it takes a relatively
+    // long time and is done directly in response to the user actions, so it
+    // makes a lot of sense to avoid doing it completely as long as possible.
+    bool must_update_columns = false;
+
+    auto const& all_header_names(case_parms()[0].member_names());
+    bool must_update_headers =
```
Comment (vadz) [inline census_view.cpp]: What's the difference between `must_update_{headers,columns}`? It looks like they end up with the same value to me...

Also, I think this `||` is too long and unwieldy, I'd use `if` for it instead, i.e.
```cpp
bool must_update_columns;
if(size changed)
{
  must_update_columns = true;
}
else
{
  must_update_columns = std::equal(...);
}
```

File: census_view.hpp [hunk 15]
```diff
@@ -134,6 +134,14 @@ class CensusView final
...
+    {
+        std::string name;
+        bool        show;
+    };
+
+    std::vector<header_cache> all_headers_;
```
Comment (vadz) [inline census_view.hpp]: Let's also rename this one to `current_headers_` to distinguish it better from `all_header_names` local variable in the code above.

**Comment by thesiv** [comment]
Fixed.

File: census_view.cpp [hunk 16]
```diff
@@ -1393,12 +1443,14 @@ void CensusView::UponColumnWidthVarying(wxCommandEvent&)
 {
     autosize_columns_ = true;
 
-    wxWindowUpdateLocker u(list_window_);
-    for(int j = 0; j < bourn_cast<int>(list_window_->GetColumnCount()); ++j)
+    if(!Update())
         {
-        list_window_->GetColumn(j)->SetWidth(wxCOL_WIDTH_AUTOSIZE);
+        wxWindowUpdateLocker u(list_window_);
+        for(unsigned int j = 0; j < list_window_->GetColumnCount(); ++j)
```
Comment (vadz) [inline census_view.cpp]: Just FYI, I've undone this (unwanted) change (as well as the similar one below) in #113.

File: any_member.hpp [hunk 17]
```diff
@@ -604,25 +603,20 @@ any_member<ClassType>& MemberSymbolTable<ClassType>::operator[]
     (std::string const& s
     )
 {
-    auto i = map_.find(s);
-    if(map_.end() == i)
+    auto i = std::find(member_names_.begin(), member_names_.end(), s);
+    if(member_names_.end() == i || s != *i)
```
Comment (vadz) [inline any_member.hpp]: It doesn't seem useful to have `|| s != *i` part here, it looks like a leftover and I'm removing it in #114, please let me know if I'm missing something here.

**Comment by vadz** [comment]
I've split this PR in #113 and #114, so closing this one.

**Comment by vadz** [comment]
Another change in #114 compared to this one: I left a perfectly stupid comment asking why `std::lower_bound()` was used instead of `find()`. The reason is, of course, that the former is `O(log N)` while the latter is `O(N)`, so it should be more efficient, and I restored its use now.

File: any_member.hpp (hunk 17 above)
Comment (thesiv) [reply any_member.hpp]: Agreed: it's unuseful.
